src/kernel/boot.c

```diff
@@ -874,20 +874,48 @@ BOOT_CODE bool_t init_freemem(word_t n_available, const p_region_t *available,
     word_t size = calculate_rootserver_size(it_v_reg, extra_bi_size_bits);
     word_t max = rootserver_max_size_bits(extra_bi_size_bits);
     for (; i >= 0; i--) {
-        word_t next = i + 1;
-        pptr_t start = ROUND_DOWN(ndks_boot.freemem[i].end - size, max);
-        if (start >= ndks_boot.freemem[i].start) {
+        /* Invariant: both i and (i + 1) are valid indices in ndks_boot.freemem. */
+        assert(i < ARRAY_SIZE(ndks_boot.freemem) - 1);
+        /* Invariant; the region at index i is the current candidate.
+         * Invariant: regions 0 up to (i - 1), if any, are additional candidates.
+         * Invariant: region (i + 1) is undefined, and should not be read.
+         * Invariant: regions above (i + 1), if any, are empty or too small to use.
+         * Invariant: defined regions (all except i + 1) are ordered,
+         *            non-overlapping and unallocated. */
+
+        /* We make a fresh variable for the unused index, because the
+         * SimplExportAndRefine verification test has poor support for array
+         * indices that are sums of variables and small constants. */
+        int unused = i + 1;
+
+        /* Can we fit a suitably aligned chunk in the top end of region i? */
+        pptr_t unaligned_start = ndks_boot.freemem[i].end - size;
+        pptr_t start = ROUND_DOWN(unaligned_start, max);
+        /* start might have underflowed, so check both ends of the region. */
+        if (start >= ndks_boot.freemem[i].start
+            && unaligned_start <= ndks_boot.freemem[i].end)
```

Comment:
I might be more intuitive is we check `unaligned_start` in both cases. And for checking the end, we should use `<` as we need some space in the slot anyway.
```suggestion
      if (unaligned_start >= ndks_boot.freemem[i].start
            && unaligned_start < ndks_boot.freemem[i].end)
```

Reply:
I'm not sure what alignment guarantees the regions in `ndks_boot.freemem` provide, but we'd need some guarantees to know that the `unaligned_start >= ndks_boot.freemem[i].start` check implies that `start >= ndks_boot.freemem[i].start`. The latter is what we really need, so I think it's clearest if we check for that. (Remember, `ROUND_DOWN` means that `start <= unaligned_start` regardless of underflow.)

For the second part, we just need to know we didn't underflow, to be sure that the first check is meaningful. And `unaligned_start <= ndks_boot.freemem[i].end` is the precise check for non-underflow. The responsibility for "we need some space" should be on `calculate_rootserver_size`. :-)

I think the problem here is the comment, so I'll clarify that. If that doesn't fix it, then we should go back to the nested conditional that you had in your PR.